shared/src/test/diff/mlscript/This.mls

```diff
@@ -0,0 +1,329 @@
+//│ Defined class DerivedDerived2
+//│ Defined DerivedDerived2.Foo2: (DerivedDerived2 & 'this) -> (DerivedDerived2 & 'this)
+
+Derived2.Foo3
+//│ res: (Derived2 & 'this) -> (Derived2 & 'this,)
+
+DerivedDerived2.Foo3
+//│ res: (DerivedDerived2 & 'this) -> (Derived2 & 'this,)
+
+// ===============
+// From Repro4.mls
+// ===============
+
+class NewBase
+  method NewFoo: this
+  method NewBar = (this.NewFoo,)
+//│ Defined class NewBase
+//│ Declared NewBase.NewFoo: (NewBase & 'this) -> (NewBase & 'this)
+//│ Defined NewBase.NewBar: (NewBase & 'this) -> (NewBase & 'this,)
+
+class NewDerived: NewBase
+  method NewQux = this.NewBar._1
+//│ Defined class NewDerived
+//│ Defined NewDerived.NewQux: (NewDerived & 'this) -> (NewDerived & 'this)
+
+// Test methods of NewDerived.
+NewDerived.NewFoo
+NewDerived.NewBar
+NewDerived.NewQux
+//│ res: (NewDerived & 'this) -> (NewBase & 'this)
+//│ res: (NewDerived & 'this) -> (NewBase & 'this,)
+//│ res: (NewDerived & 'this) -> (NewDerived & 'this)
+
+class NewDerivedDerived: NewDerived
+  method NewQuz = (this.NewQux)
+//│ Defined class NewDerivedDerived
+//│ Defined NewDerivedDerived.NewQuz: (NewDerivedDerived & 'this) -> (NewDerivedDerived & 'this)
+
+NewDerivedDerived.NewFoo
+NewDerivedDerived.NewBar
+NewDerivedDerived.NewQux
+NewDerivedDerived.NewQuz
+//│ res: (NewDerivedDerived & 'this) -> (NewBase & 'this)
+//│ res: (NewDerivedDerived & 'this) -> (NewBase & 'this,)
+//│ res: (NewDerivedDerived & 'this) -> (NewDerived & 'this)
+//│ res: (NewDerivedDerived & 'this) -> (NewDerivedDerived & 'this)
+
+:e
+class NewDerivedDerivedDerived: NewDerivedDerived
+  method NewQux = this.NewBar._1
+//│ ╔══[ERROR] Overriding method NewDerived.NewQux without explicit declaration is not allowed.
+//│ ║  l.293: 	  method NewQux = this.NewBar._1
+//│ ║         	         ^^^^^^^^^^^^^^^^^^^^^^^
+//│ ╟── Note: method definition inherited from
+//│ ║  l.265: 	  method NewQux = this.NewBar._1
+//│ ╙──       	         ^^^^^^^^^^^^^^^^^^^^^^^
+//│ Defined class NewDerivedDerivedDerived
+//│ Defined NewDerivedDerivedDerived.NewQux: (NewDerivedDerivedDerived & 'this) -> (NewDerivedDerivedDerived & 'this)
+
+// Test methods of NewDerivedDerivedDerived
+NewDerivedDerivedDerived.NewBar
+NewDerivedDerivedDerived.NewFoo
+NewDerivedDerivedDerived.NewQuz
+NewDerivedDerivedDerived.NewQux
+//│ res: (NewDerivedDerivedDerived & 'this) -> (NewBase & 'this,)
+//│ res: (NewDerivedDerivedDerived & 'this) -> (NewBase & 'this)
+//│ res: (NewDerivedDerivedDerived & 'this) -> (NewDerivedDerived & 'this)
+//│ res: (NewDerivedDerivedDerived & 'this) -> (NewDerivedDerivedDerived & 'this)
+
+// ===============
+// From Repro5.mls
+// ===============
+
+class Animal
+  method Eat = this
+//│ Defined class Animal
+//│ Defined Animal.Eat: (Animal & 'this) -> (Animal & 'this)
+
+class Cat: Animal
+//│ Defined class Cat
+
+Animal.Eat
+//│ res: (Animal & 'this) -> (Animal & 'this)
+
+Cat.Eat
+//│ res: (Cat & 'this) -> (Animal & 'this)
```

Comment:
This test file is full of garbage, like stuff copied from other tests that doesn't have anything to do with the `this`-type feature... I had to go through it and clean it up myself.